JDK-8328190 : Convert AWTPanelSmoothWheel.html applet test to main

honkar-jdk opened the pull request:
This test is converted to main using PassFailJFrame. It verifies wheel rotation value for high-res mouse on windows.

The test requires the updated PassFailJFrame changes added in this PR https://github.com/openjdk/jdk/pull/18281

honkar-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,84 @@
+    public static final String INSTRUCTIONS = """
+            This test is relevant for windows platforms and mouses with high-resolution wheel,
+            please just press pass if this is not the case.
+            
```
@prrace I wanted to check the relevance of the testing requirement. The original test was created for Windows Vista and for a high-res mouse. I do think the test can be retained to check on newer versions of windows. Any suggestions?

aivanov-jdk replied:
A trackpad is usually a high-res mouse wheel, it allows smooth scrolling (as opposed to moving 3 lines on a wheel notch).

There are mice which also support smooth scrolling, I've got one.

honkar-jdk commented:
This PR has been updated to use PassFailJFrame's new logArea() feature.
@azvegint @aivanov-jdk Please review.

turbanoff commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,76 @@
+        Panel panel = new Panel();
+        panel.setBackground(Color.GREEN);
+        panel.addMouseWheelListener(e -> PassFailJFrame.log(e.toString()));
+        frame.setSize (200,200);
```
```suggestion
        frame.setSize(200, 200);
```

azvegint commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,76 @@
+            """;
+
+    public static void main (String[] args) throws Exception {
+        PassFailJFrame.builder()
```
It can be added, but I see no reason to do so.

It may be helpful for tests that can be run standalone (that don't respect the jtreg tags and depend on some other jtreg libraries), but for this one the only easy way to run it is through jtreg, so the `@requires (os.family == "windows")` is sufficient.

honkar-jdk replied:
I agree with @azvegint, since it has @requires check in jtreg and the test will mainly be run using jtreg framework , the OS check is not strictly required.

kumarabhi006 replied:
Fair enough.

aivanov-jdk replied:
I agree. Adding an explicit OS check adds one more dependency on a test library.

It's not that hard to run a test with `PassFailJFrame` without jtreg, however, it's not as simple as `java <path>/<testname>.java`. Any way, if you run on a non-Windows, you may get unexpected result.

Is the test not applicable to other platforms? macOS also supports high-res mice, trackpad on a MacBook is high-res too.

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,76 @@
+/*
+ * @test
+ * @bug 6730447
+ * @summary [Win] To verify the support for high resolution mouse wheel on Windows.
```
This could be applicable to other platforms too, even though [JDK-6730447](https://bugs.openjdk.org/browse/JDK-6730447) and [JDK-6524352](https://bugs.openjdk.org/browse/JDK-6524352) are reported against Windows only. It could be than support for high-resolution mouse wheel was implemented on macOS earlier than on Windows.

honkar-jdk replied:
I have remove @requires tag and the test is extended to all platforms.

honkar-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,76 @@
+            Place the mouse cursor above the green panel and rotate the mouse wheel,
+            the test will print all mouse wheel event messages into the logging panel
+            below the instruction window.
+            Please make sure that some of the messages have non-zero 'wheelRotation' value,
+            and also check if the test works OK if the mouse wheel is rotated very slow.
```
Sounds good, test can be made semi-automated test.
How long or many MouseWheelEvents do we check ? Probably 10-15 would be a good number?

honkar-jdk replied:
@aivanov-jdk Semi-automated the test, it now passes when 5 or more  MouseWheekEvent of type: scrollType=WHEEL_UNIT_SCROLL & wheelRotation != 0 is recorded. Updated the instructions accordingly.

aivanov-jdk replied:
I think we should update the instructions.

A high resolution mouse produces events like this:
```text
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(164,90),absolute(1135,365),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=0,preciseWheelRotation=0.25] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(164,90),absolute(1135,365),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=0,preciseWheelRotation=0.25] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(164,90),absolute(1135,365),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=0,preciseWheelRotation=0.25] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(133,46),absolute(1104,321),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=1,preciseWheelRotation=1.0] on panel0
```

Thus, `wheelRotation=0` while `preciseWheelRotation=0.25`. When it reaches, `preciseWheelRotation=1.0`, the event contains **`wheelRotation=1`**.

A regular mouse with mouse-wheel notches produces the following events:
```text
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(90,115),absolute(1061,390),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=1,preciseWheelRotation=1.0] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(176,76),absolute(1147,351),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=1,preciseWheelRotation=1.0] on panel0
```

As you can see, `scrollType=WHEEL_UNIT_SCROLL` is the same in both cases. The only difference is that high-resolution mouse produces events where **`preciseWheelRotation < 1.0`**.

This means that you should also track the value of `preciseWheelRotation`. Otherwise, the test will pass when a regular mouse is used.

You may show a warning to the user if you detect, let's say, 5 events with `wheelRotation=1` but none of the events had `wheelRotation=0` and `preciseWheelRotation < 1.0`.

honkar-jdk replied:
I believe since there a lot of conditions that needs to be checked manually, it is better to convert this test from semi-automatic to completely manual.

I can parse the MouseWheelEvent to display only the info needed by the user such as PreciseWheelRotation and WheelRotation for easier checking and include a warning when not using a hi-res mouse.

honkar-jdk replied:
@aivanov-jdk Thanks for reviewing this in-depth and pointing out the difference between wheelRotation & preciseWheelRotation when using hi-res mouse.

honkar-jdk replied:
@aivanov-jdk Since preciseWheelRotation amount depends on user's movement of the high-res mouse and there might be a need to look through logs for few events, it is better to convert this test from semi-automated to manual. Does this sound good. Updated the PR with the latest changes.

aivanov-jdk replied:
@honkar-jdk I think the test could still remain semi-automatic. Can the test verify the conditions? I see no reason why it can't… yet it requires more code. It can be postponed to a later time, if necessary.

I don't think we should make the tester read the log and verify the conditions which can be verified automatically.

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -25,50 +25,82 @@
 
-            If the above is true press PASS, else FAIL.
+            A hi-res mouse is one which produces MouseWheelEvents having
+            <b>preciseWheelRotation &lt 1.</b> <br>
+            When preciseWheelRotation adds up to 1,wheelRotation becomes 1. <br>
+            You should see a few events where preciseWheelRotation < 1 & wheelRotation = 0
+            followed by a event where preciseWheelRotation = 1 & wheelRotation = 1.<br> <br>
```
```suggestion
            <b>preciseWheelRotation &lt; 1.</b> <br>
            When preciseWheelRotation adds up to 1, wheelRotation becomes 1. <br>
            You should see a few events where preciseWheelRotation &lt; 1 &amp; wheelRotation = 0
            followed by a event where preciseWheelRotation = 1 &amp; wheelRotation = 1.<br> <br>
```

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -25,50 +25,82 @@
+            Please press PASS if above is true, else FAIL. <br> <br>
+
+            <hr>
+            PS: If you don't see events with preciseWheelRotation < 1,
```
```suggestion
            PS: If you don't see events with preciseWheelRotation &lt; 1,
```

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -25,50 +25,82 @@
-        frame.setSize (200, 200);
+        panel.addMouseWheelListener(e -> {
+            if (e.getScrollType() == MouseWheelEvent.WHEEL_UNIT_SCROLL) {
+                PassFailJFrame.log("WheelEvent#"+ ++wheelEventCount
```
```suggestion
                PassFailJFrame.log("WheelEvent#" + (++wheelEventCount)
```
Use parentheses to avoid confusion.

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -25,50 +25,82 @@
+                }
+            }
+        });
+        frame.setSize (400, 200);
```
```suggestion
        frame.setSize(400, 200);
```

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -25,50 +25,82 @@
+                if (wheelEventCount >= 5 && hiResWheelCount == 0) {
+                    PassFailJFrame.log("WARNING !!! You might not be using a high-res mouse.");
+                }
+            }
```
This logging style is easier to parse, thanks.

You have nearly all the data to make the test semi-automatic. What's missing is `wheelRotationCount`.

[This diff](https://github.com/aivanov-jdk/jdk/commit/04c80c519d1c116663ab10a177bff1c4b6924091) implements this feature. For convenience, I'll paste the diff inline too:
```diff
diff --git a/test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java b/test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java
index 1097d1dcb04..f3dd8294e8b 100644
--- a/test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java
+++ b/test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java
@@ -39,7 +39,9 @@

 public class AWTPanelSmoothWheel {
     private static int wheelEventCount = 0;
+    private static int wheelRotationCount = 0;
     private static int hiResWheelCount = 0;
+
     private static final String INSTRUCTIONS = """
             <html>
             <body>
@@ -92,12 +94,21 @@ private static Frame createUI() {
                         + " --- Wheel Rotation: " + e.getWheelRotation()
                         + " --- Precise Wheel Rotation: "
                         + String.format("%.2f", e.getPreciseWheelRotation()));
+                if (e.getWheelRotation() >= 1) {
+                    wheelRotationCount += e.getWheelRotation();
+                }
                 if (e.getPreciseWheelRotation() < 1) {
                     hiResWheelCount++;
                 }
                 if (wheelEventCount >= 5 && hiResWheelCount == 0) {
                     PassFailJFrame.log("WARNING !!! You might not be using a high-res mouse.");
                 }
+                if (wheelRotationCount > 5
+                    && (hiResWheelCount / 2 >= wheelRotationCount)) {
+                    PassFailJFrame.log("The test passes: hiResWheelCount = " + hiResWheelCount
+                                       + " wheelRotationCount = " + wheelRotationCount);
+                    PassFailJFrame.forcePass();
+                }
             }
         });
         frame.setSize (400, 200);
```

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,108 @@
+                    hiResWheelCount++;
+                }
+                if (wheelEventCount >= 5 && hiResWheelCount == 0) {
+                    PassFailJFrame.log("WARNING !!! You might not be using a high-res mouse.");
```
[`JOptionPane.showMessageDialog`](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/javax/swing/JOptionPane.html#showMessageDialog(java.awt.Component,java.lang.Object,java.lang.String,int)) could be a better option to attract the tester's attention.

honkar-jdk replied:
Thanks for the suggestion. Added it.

honkar-jdk commented:
@aivanov-jdk 

I have changed the logic for MouseWheelEvents as below:

- wheelRotationCount was not required, removed it .
- checking for Math.abs(e.getPreciseWheelRotation()) value since it can be negative when mouse is scrolled down.
- `(e.getWheelRotation() != 0 && hiResWheelCount > 0)` this condition looks to be sufficient to forcePass the test as we need only one complete non-zero wheelRotation value.
- added JOptionPane for the warning message as suggested.

aivanov-jdk commented:
I somewhat dislike that the test completes too quickly when I use my mouse. Yes, you warn the tester the test is semi-automatic, yet I'd prefer giving the tester more time.

Technically, one event with `e.getWheelRotation() != 0` is enough, that's right. It just doesn't feel right…

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,132 @@
+            preciseWheelRotation &lt; 1 &amp; wheelRotation = 0 followed by
+            an event where wheelRotation != 0 in the logs.<br> <br>
+
+            Check if the test works OK when the mouse wheel is rotated very slow.<br> <br>
```
```suggestion
            Check if the test works OK when the mouse wheel is rotated very slowly.<br> <br>
```

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,132 @@
+            an event where wheelRotation != 0 in the logs.<br> <br>
+
+            Check if the test works OK when the mouse wheel is rotated very slow.<br> <br>
+            This is a semi-automated test, if you are using a hi-res mouse and
```
I would move this paragraph above. At this moment, it's more important to say that the test will pass automatically than that the tester needs to look through the log of events. So, the sentence above could be removed altogether.

honkar-jdk replied:
Updated.

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,132 @@
+            the test should automatically pass.<br> <br>
+
+            <hr>
+            PLEASE NOTE:
```
```suggestion
            <b>Note:</b>
```
You have formatting options, bold text will stand out and it's usually easier to read than uppercase text.

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,132 @@
+                <li> When mouse is scrolled up, preciseWheelRotation & wheelRotation are positive
+                 and they are negative when scrolled down. </li>
+            </ul>
+            <br>
```
```suggestion
```
May be removed.

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,132 @@
+                if (wheelEventCount >= 5 && hiResWheelCount == 0) {
+                    PassFailJFrame.log(WARNING_MSG);
+                    JOptionPane.showMessageDialog(frame, WARNING_MSG,
+                            "WARNING", WARNING_MESSAGE);
```
```suggestion
                            "WARNING", JOptionPane.WARNING_MESSAGE);
```
I suggest using `JOptionPane.WARNING_MESSAGE` explicitly: there's no confusion between `WARNING_MSG` and `WARNING_MESSAGE` then.

I'd use ‘Warning’ in its title case as opposed to uppercase, taking into account the message itself contains upper case warning.

aivanov-jdk commented on test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java:
```diff
@@ -0,0 +1,132 @@
+    private static final String INSTRUCTIONS = """
+            <html>
+            <body>
+            This test is relevant on platforms with high-resolution mouse wheel,
+            please press PASS if this is not the case.<br> <br>
```
It may be worth mentioning that trackpad will suit too.

honkar-jdk replied:
Updated.

aivanov-jdk reviewed (approved):
It looks good to me.

honkar-jdk commented:
/integrate

aivanov-jdk commented:
> I somewhat dislike that the test completes too quickly when I use my mouse. Yes, you warn the tester the test is semi-automatic, yet I'd prefer giving the tester more time.
> 
> Technically, one event with `e.getWheelRotation() != 0` is enough, that's right. It just doesn't feel right…

I run the test it with trackpad on my Windows laptop. Indeed, waiting for `e.getWheelRotation() != 0` takes too long if `e.getPreciseWheelRotation()` is very small, it is 0.01 on Harshitha's laptop and on mine too.

When the value of `e.getPreciseWheelRotation() > 0.10`, it feels the test completes too fast.

We can change the sensitivity of the test. Is it worth the effort? Perhaps, it isn't.

Does anyone else have their preferences?
